opencensus/trace/propagation/google_cloud_format.py

```diff
@@ -19,7 +19,8 @@
 from opencensus.trace.trace_options import TraceOptions
 
 _TRACE_CONTEXT_HEADER_NAME = 'X-Cloud-Trace-Context'
-_TRACE_CONTEXT_HEADER_FORMAT = r'([0-9a-f]{32})(\/([0-9a-f]{16}))?(;o=(\d+))?'
+_TRACE_CONTEXT_HEADER_FORMAT = r'([0-9a-f]{32})'\
+    r'(\/([0-9a-f]{16}|(None|)))?(;o=(\d+))?'
```

Comment:
It makes more sense to use `0000000000000000` than `None` unless:
1. `None` is a well documented format for `X-Cloud-Trace-Context`, here we're just trying to follow the spec.
2. We need to support `None` for backward compatibility (as there are existing services which send `None` already, and we need to interop with them).

Reply:
Regarding 2; the old behaviour wrongly parsed the trace options but silently ignored the wrong span_id. With this new regex there is still span_id=None, but a warning is logged and the trace options are correct. See a header being parsed with the old and the new function below:

```python

In [131]: header3_on                                                                                                                                                                                                      
Out[131]: '6e0c63257de34c92bf9efcd03927272e/None;o=1'

In [132]: orig_from_header(header3_on)                                                                                                                                                                                    
Out[132]: SpanContext(trace_id=6e0c63257de34c92bf9efcd03927272e, span_id=None, trace_options=TraceOptions(enabled=True), tracestate=None)

In [133]: new_from_header(header3_on)                                                                                                                                                                                     
WARNING:root:Span_id None does not the match the required format
Out[133]: SpanContext(trace_id=6e0c63257de34c92bf9efcd03927272e, span_id=None, trace_options=TraceOptions(enabled=True), tracestate=None)
```

```python
In [134]: header3_off                                                                                                                                                                                                     
Out[134]: '6e0c63257de34c92bf9efcd03927272e/None;o=0'

In [135]: orig_from_header(header3_off)                                                                                                                                                                                   
Out[135]: SpanContext(trace_id=6e0c63257de34c92bf9efcd03927272e, span_id=None, trace_options=TraceOptions(enabled=True), tracestate=None)

In [136]: new_from_header(header3_off)                                                                                                                                                                                    
WARNING:root:Span_id None does not the match the required format
Out[136]: SpanContext(trace_id=6e0c63257de34c92bf9efcd03927272e, span_id=None, trace_options=TraceOptions(enabled=False), tracestate=None)
```

Reply:
Do you think it's still worth handling the "None" case here now that `to_header` is fixed? One argument in favor is that this will work with older versions of the library that included the bug, but we wouldn't check for "None" here if the bug had never existed.

Reply:
I'm happy to remove it, the argument would indeed be backwards compatibility with 'old' incoming headers, but seeing as the current behaviour isn't correct anyways, and it's only a small and non-breaking difference in behaviour, I would agree that it might not be worth including it.

Related, what's your opinion on handling headers with no span id at all (so e.g. `6e0c63257de34c92bf9efcd03927272e/;o=0` or without the slash `6e0c63257de34c92bf9efcd03927272e;o=0`)? If that should not be handled, then the extra group can be removed from the regex. To support it and lt resulting in a spancontext without span_id but with the right trace_options, an empty group can be kept where now the `None|` is.